[SPARK-38887][SQL] Support switch inner join side for sort merge join

ulysses-you opened the pull request:
<!--
Thanks for sending a pull request!  Here are some tips for you:
  1. If this is your first time, please read our contributor guidelines: https://spark.apache.org/contributing.html
  2. Ensure you have added or run the appropriate tests for your PR: https://spark.apache.org/developer-tools.html
  3. If the PR is unfinished, add '[WIP]' in your PR title, e.g., '[WIP][SPARK-XXXX] Your PR title ...'.
  4. Be sure to keep the PR description updated to reflect all changes.
  5. Please write your PR title to summarize what this PR proposes.
  6. If possible, provide a concise example to reproduce the issue for a faster review.
  7. If you want to add a new configuration, please read the guideline first for naming configurations in
     'core/src/main/scala/org/apache/spark/internal/config/ConfigEntry.scala'.
  8. If you want to add or modify an error type or message, please read the guideline first in
     'core/src/main/resources/error/README.md'.
-->

### What changes were proposed in this pull request?
<!--
Please clarify what changes you are proposing. The purpose of this section is to outline the changes and how this PR fixes the issue. 
If possible, please consider writing useful notes for better and faster reviews in your PR. See the examples below.
  1. If you refactor some codes with changing classes, showing the class hierarchy will help reviewers.
  2. If you fix some SQL features, you can provide some references of other DBMSes.
  3. If there is design documentation, please add the link.
  4. If there is a discussion in the mailing list, please add the link.
-->
Enhance the selection of `SortMergeJoin` streamed side and buffered side in `JoinSelection`, so that we can get a buffered side for better performance. 

### Why are the changes needed?
<!--
Please clarify why the changes are needed. For instance,
  1. If you propose a new API, clarify the use case for a new API.
  2. If you fix a bug, you can clarify why it is a bug.
-->
For an inner join type SortMergeJoin, it always uses the left side as streamed side and right side as buffered side.

Accoirding to the implementaion of SortMergeJoin, we expect the buffered side to be:

- smaller than streamed side
- has less duplicate data

We do not know whether the join will be SortMergeJoin at logical phase, so it should do this selection at physcial phase.

For example:
```sql
SELECT * FROM (
  SELECT c1 FROM a GROUP BY c1 
) t1 JOIN t2 ON t1.c1 = t2.c2
```
We can make t1 as the buffered side if its statistics size less than t2.

### Does this PR introduce _any_ user-facing change?
<!--
Note that it means *any* user-facing change including all aspects such as the documentation fix.
If yes, please clarify the previous behavior and the change this PR proposes - provide the console output, description and/or an example to show the behavior difference if possible.
If possible, please also clarify if this is a user-facing change compared to the released Spark versions or within the unreleased branches such as master.
If no, write 'No'.
-->
The plan may be changed and improve the performance.

### How was this patch tested?
<!--
If tests were added, say they were added here. Please make sure to add some test cases that check the changes thoroughly including negative and positive cases if possible.
If it was tested in a way different from regular unit tests, please clarify how you tested step by step, ideally copy and paste-able, so that other reviewers can test and check, and descendants can verify in the future.
If tests were not added, please describe why they were not added and/or why it was difficult to add.
If benchmark tests were added, please run the benchmarks in GitHub Actions for the consistent environment, and the instructions could accord to: https://spark.apache.org/developer-tools.html#github-workflow-benchmarks.
-->
add test

sigmod commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkStrategies.scala:
```diff
@@ -236,8 +236,10 @@ abstract class SparkStrategies extends QueryPlanner[SparkPlan] {
-              leftKeys, rightKeys, joinType, nonEquiCond, planLater(left), planLater(right))))
+            val (streamed, buffered) = getSortMergeJoinStreamedAndBufferedSide(joinType, leftKeys,
+              left, right, conf)
+            Some(Seq(joins.SortMergeJoinExec(leftKeys, rightKeys, joinType, nonEquiCond,
```
Just curious: does the order of output attributes matter?

https://github.com/apache/spark/blob/d8920ae13346670bac3f1a7e49dfc2ba85c93b03/sql/core/src/main/scala/org/apache/spark/sql/execution/joins/ShuffledJoin.scala#L62

iiuc, BHJ has an BuildSide enum and doesn't actually switch left/right children so it doesn't have this problem?

ulysses-you replied:
yeah, I think we should change it back to the original smj output ordering after we switch join side.

ulysses-you commented:
Some performance number with the benchmark:

```
 sort merge join with buffered side duplicates, switched: true,:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 -------------------------------------------------------------------------------------------------------------------------------------------------------------
 sort merge join with buffered side duplicates, switched: true, wholestage off           5175           5196          30          6.5         154.2       1.0X
 sort merge join with buffered side duplicates, switched: true, wholestage on            4991           5221         407          6.7         148.7       1.0X
 Running benchmark: sort merge join with buffered side duplicates, switched: false,
   Running case: sort merge join with buffered side duplicates, switched: false, wholestage off
   Stopped after 2 iterations, 12610 ms
   Running case: sort merge join with buffered side duplicates, switched: false, wholestage on
   Stopped after 5 iterations, 33804 ms
 OpenJDK 64-Bit Server VM 1.8.0_322-b06 on Mac OS X 12.3.1
 Apple M1 Pro
 sort merge join with buffered side duplicates, switched: false,:                Best Time(ms)   Avg Time(ms)   Stdev(ms)    Rate(M/s)   Per Row(ns)   Relative
 --------------------------------------------------------------------------------------------------------------------------------------------------------------
 sort merge join with buffered side duplicates, switched: false, wholestage off           6291           6305          20          5.3         187.5       1.0X
 sort merge join with buffered side duplicates, switched: false, wholestage on            6635           6761         126          5.1         197.7       0.9X
```

ulysses-you committed:
refactor
